pkg/applicationmanager/controllers/applicationbackup.go

```diff
@@ -730,39 +729,18 @@ func (a *ApplicationBackupController) backupVolumes(backup *stork_api.Applicatio
 					// process is considered to be completed successfully.
 					// This ensures we don't execute the post-exec before all volume's snapshot is completed
 					if driverName == volume.PortworxDriverName {
-						startTime := time.Now()
-						for {
-							// In case below logic genuinely takes long time for many volumes
-							// the CR timestamp need to be updated. This prevents backup timeout error.
-							elapsedTime := time.Since(startTime)
-							if elapsedTime > utils.TimeoutUpdateBackupCrTimestamp {
-								backup, err = a.updateBackupCRInVolumeStage(
-									namespacedName,
-									stork_api.ApplicationBackupStatusInProgress,
-									backup.Status.Stage,
-									"Volume backups are in progress",
-									volumeInfos,
-								)
-								if err != nil {
-									return err
-								}
-								startTime = time.Now()
-							}
-							// Get fresh stock of the status for all volumes.
-							snapshotNotCompleted := false
-							volumeInfos, err = driver.GetBackupStatus(backup)
-							if err != nil {
-								log.ApplicationBackupLog(backup).Errorf("error getting backup status for driver %v: %v", driverName, err)
-								return fmt.Errorf("error getting backup status for driver %v: %v", driverName, err)
-							}
-							for _, volInfo := range volumeInfos {
-								if volInfo.BackupID == "" {
-									log.ApplicationBackupLog(backup).Tracef("Snapshot of volume %v is not done yet, need to loop till snapshot is done...", volInfo.PersistentVolumeClaim)
-									snapshotNotCompleted = true
-								}
-							}
-							if !snapshotNotCompleted {
-								break
+						volumeInfos, err = driver.GetBackupStatus(backup)
+						if err != nil {
+							return fmt.Errorf("error getting backup status: %v", err)
+						}
+						for _, volInfo := range volumeInfos {
+							if volInfo.BackupID == "" {
+								log.ApplicationBackupLog(backup).Infof("Snapshot of volume [%v] hasn't completed yet, let's retry snapshot", volInfo.PersistentVolumeClaim)
+								// Some volume snapshot is not completed yet
+								// hence we will retry checking the status in the next reconciler iteration
+								// *stork_api.ApplicationBackupVolumeInfo.Status is not being checked here
+								// since backpID confirms if the snapshot is done or not already
+								return nil
 							}
 						}
 					}
```

Comment:
With the current logic, we are starting backup for only one batch of PVC in a reconciler cycle. Should the checking of the snapshot ID be outside the for loop, where startbackup and before the code that starts Postexec rule.
With the current, we are unnecessarily serialising the backup start, when we have more number of PVC. Please check it out.

Reply:
I am bit confused here.. 
1. In 1st for loop we are iterating over all volumes eligible to be backed up... 
2. In 2nd for loop we are invoking startBackup()  in batches  till we  completely call this for all the PVCs  in batches
3. And in this 2nd for loop, per iteration I have added another for loop to run through all the PVCs in that batch to check if snapshot is done or not.  

- failure scenario In the 2nd for loop itself If any pvc is not backed up yet then return and let the reconciler kick in again to start from the PVC wherein we have left, by collecting all the PVCs freshly which is not backed up yet.
- success scenario - And in case that batch is completed gracefully we will go for 2nd batch again till we finish all  pvcs. If Px backend respond success for every batch of pvcs then in a single reconciler call itself we will finsh backup of all PVCs. 

I am unable to conceive without a failure how the  for-loop introduced has made each reconciler loop only takes care of one batch of volume.  This will happen only if in a particular batch volume backup hasn't completed and we need to return from reconciler. And in case we identified backup is not completed at backend it make sense to come back again in reconciler other wise we will start executing post-exec.

Reply:
@siva-portworx  I see your point here. every time a batch is picked to be backed-up and first backup is not set  we will return from reconciler, then by that time that batch may have completed and the same thing can repeat for 2nd batch that way every reconciler  loop we will consider one one batch only in worst case scenario.

So as you suggested let me put this check out side of for loops so that the check happen after we issue backup for all the batches then a reconciler based wait makes it more faster..